R/time_bins.R

```diff
@@ -293,74 +304,124 @@ time_bins <- function(interval = "Phanerozoic", rank = "stage", size = NULL,
+      for (i in k:n_intervals) {  # Intervals from k to n_intervals
+        for (j in (k - 1):(i - 1)) {  # Possible partition points
+          bin_duration <- cumulative_durations[i + 1] - cumulative_durations[j + 1]
+          cost <- M[j + 1, k - 1] + (bin_duration - target_bin_size)^2
+          if (cost < M[i + 1, k]) {
+            M[i + 1, k] <- cost
+            partition[i + 1, k] <- j
+          }
+        }
       }
-      # Generate dataframe
-      grouping_rank <- rank
-      df <- cbind.data.frame(bin,
-                             max_ma,
-                             mid_ma,
-                             min_ma,
-                             duration_myr,
-                             grouping_rank,
-                             intervals)
-      # Message user
-      message(
-        paste("Target equal length time bins was set to",
-              round(size, digits = 2),
-              "Myr. \nGenerated time bins have a mean length of",
-              round(mean(df$duration_myr), digits = 2),
-              "Myr and a standard deviation of",
-              round(sd(df$duration_myr), digits = 2),
-              "Myr."
-        )
+    }
+    #
+    # Backtracking to find the optimal partitioning
+    bins <- vector("list", n_bins) # empty list for bins
+    i <- n_intervals
+    k <- n_bins
+    while (k > 0) {
+      j <- partition[i + 1, k]
+      bins[[k]] <- (j + 1):i
+      i <- j
+      k <- k - 1
+    }
+    } else {
+      # if only 1 bin
+      bins <- list(seq_len(nrow(df)))
+    }
+    #
+    # Assemble bins and compute bin statistics
+    # Initialise an empty list to store the rows for bin_data
+    binDataList <- vector("list", length(bins))
+    # Iterate over bins
+    for (i in seq_along(bins)) {
+      binIntervals <- bins[[i]]
+      binMaxMa <- max(df[binIntervals, "max_ma"])
+      binMinMa <- min(df[binIntervals, "min_ma"])
+      # Collect data for each bin in a list
+      binDataList[[i]] <- data.frame(
+        bin = as.integer(i),
+        max_ma = as.numeric(binMaxMa),
+        mid_ma = as.numeric((binMinMa + binMaxMa) / 2),
+        min_ma = as.numeric(binMinMa),
+        duration_myr = as.numeric(sum(df[binIntervals, "duration_myr"])),
+        grouping_rank = rank,
+        intervals = toString(df[binIntervals, "interval_name"]),
+        stringsAsFactors = FALSE
       )
+    }
+    # List to dataframe
+    df <- do.call(rbind, binDataList)
+    #
+    # Calculate mean and standard deviation of bin durations
+    mean_duration <- mean(df$duration_myr)
+    sd_duration <- sd(df$duration_myr)
+    #
+    # Message user
+    if (is.numeric(size)) {
+      message_head <- paste("Target duration of equal length time bins was set to",
+                            round(size, digits = 2),  "Myr.\n")
+    } else {
+      message_head <- paste0("Number of equal length time bins was set to ",
+                            n_bins, ".\n")
+    }
+    message(
+      paste(message_head,
```

Comment:
Should change to using paste0 here, otherwise we get a space before the "time bins were generated" line

Reply:
Implemented